### PR TITLE
Drop java-10-openjdk

### PR DESCRIPTION
**Pull request by mloviska**
- Related ticket: [https://progress.opensuse.org/issues/64532](https://progress.opensuse.org/issues/64532)
- Verification runs:
  * [sle-15-Server-DVD-Updates-x86_64](https://openqa.suse.de/tests/4018117)
  * [sle-15-SP1-Server-DVD-Updates-x86_64](https://openqa.suse.de/tests/4018114)
  * [sle-15-SP2-JeOS-for-kvm-and-xen-x86_64](https://openqa.suse.de/tests/4018113)
  * [sle-12-SP4-Server-DVD-Updates-x86_64](https://openqa.suse.de/tests/4018115)
  * [sle-12-SP5-Server-DVD-Updates-x86_64](https://openqa.suse.de/tests/4018116)
  * [opensuse-15.1-JeOS-for-kvm-and-xen-x86_64](https://openqa.opensuse.org/tests/1209326)
  * [opensuse-15.2-JeOS-for-kvm-and-xen-x86_64](https://openqa.opensuse.org/tests/1209327)
  * [sle-15-SP2-Online-x86_64](https://openqa.suse.de/t4019023)

* `java-*bootstrap` support has been removed.
* soft fail code for bsc#1137466 has been removed since java-10 comes from legacy module

### CODE REVIEW

File: schedule/test.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+schedule:
```
Comment (Vogtinator) [inline schedule/test.yaml]: Those files shouldn't be part of this PR I guess?

File: tests/console/java.pm [hunk 2]
```diff
@@ -21,56 +21,44 @@ use base "consoletest";
...
-            diag "There are java bootstrap packages available to be installed";
-            print "There are java bootstrap packages available to be installed\n";
-            if ($bootstrap_conflicts_rt == 0) {
-                diag "There is no conflict installing the java bootstrap packages";
-                print "There is no conflict installing the java bootstrap packages\n";
-                zypper_call("in java-*");
+    if (my $rc = zypper_call('in --auto-agree-with-licenses java-*', exitcode => [0, 4, 107])) {
+        # Workaround bsc#1115375, allow script failures
+        if ($rc == 107) {
+            unless (script_run('grep "java-9-openjdk-javadoc.*scriptlet failed" /var/log/zypp/history')) {
```
Comment (Vogtinator) [inline tests/console/java.pm]: This would also break if some other scriptlet fails in addition. I'd just keep the fail and not try to detect the softfail for this..

Comment (mloviska) [reply tests/console/java.pm]: the `else` belongs to this `unless`. So we should be covered, mainly the whole handling of 107 is part of this softfail code. When everything works fine, we should handle only the error code 4 which represents softfail [https://bugzilla.suse.com/show_bug.cgi?id=1137466](Bug 1137466 - [Build 20190606-1] sle15sp{0,1} pull obsoleted java-10 packages).

**Comment by mloviska** [comment]
@Vogtinator I have also dropped the handling of `java-*bootstrap` in opensuse. Any idea what happened to these packages? I couldn't find them in OpenSUSE.

**Comment by Vogtinator** [comment]
> @Vogtinator I have also dropped the handling of `java-*bootstrap` in opensuse. Any idea what happened to these packages? I couldn't find them in OpenSUSE.

Those were only used inside OBS and never meant to be published. I assume there were some old builds where -bootstrap packages ended up by mistake, this was working around that.

File: tests/console/java.pm (hunk 2 above)
Comment (Vogtinator) [reply tests/console/java.pm]: > the else belongs to this unless.

Yes, but `grep` succeeds if the scriptlet failed and doesn't look at other failures.

> When everything works fine, we should handle only the error code 4

`4` means `ZYPPER_EXIT_ERR_ZYPP` and must not be taken as meaning success in any case.

The proper way for testing this is to explicitly exclude the obsolete JRE 10.

Comment (mloviska) [reply tests/console/java.pm]: soft fail only in case of 4 or 107 returned from zypper and each of them has to match specific recorded failure

File: tests/console/java.pm [hunk 3]
```diff
@@ -21,56 +21,46 @@ use base "consoletest";
...
-                zypper_call "in `(zypper se java-* | grep -v bootstrap | grep -v 'i ' | awk '{print \$2}' | sed -n -E -e '/java/,\$ p')`";
-            }
-        }
-        else {
-            diag "There are no java bootstrap packages";
-            print "There are no java bootstrap packages\n";
-            zypper_call("in java-*");
+        # Workaround#2 java-10 is availabe, despite java-11 obsoletes it:
+        # install java-11-openjdk* & java-1_* in opensuse & sle
+        # additionally install java-*-ibm* is sle
```
Comment (Vogtinator) [inline tests/console/java.pm]: `in sle`?

File: tests/console/java.pm (hunk 2 above)
Comment (Vogtinator) [reply tests/console/java.pm]: The match for `107` isn't specific enough though. If the log contains a mention of the grepped line anywhere, it accepts it.

IMO a softfail for this is just not worth it.

For the `4` case it's ok, because there's an explicit retry.

Comment (mloviska) [reply tests/console/java.pm]: > The match for `107` isn't specific enough though. If the log contains a mention of the grepped line anywhere, it accepts it.
> 
> IMO a softfail for this is just not worth it.
> 
I assume ( I have grepped tests and lib folders in os-autoinst-distri-opensuse )  that `java.pm` is the only intended java greedy installer. Otherwise `zypper in java` should download java-11. Maybe I am too much optimistic, but it shouldn't be pulled anywhere else. Maybe the package should be even deprecated since I haven't noticed it in sle12 during VRs.

Comment (mloviska) [reply tests/console/java.pm]: and with respect to https://bugzilla.suse.com/show_bug.cgi?id=1137466 claims that java-10 is obsoleted.

Comment (Vogtinator) [reply tests/console/java.pm]: Yes, but my point is that if later other java scriptlets also start to fail in addition to `java-9-openjdk-javadoc`, that would go entirely unnoticed. This scenario isn't too unlikely, as fixes get lost and regressions appear quite often.

Comment (mloviska) [reply tests/console/java.pm]: Yup and then it should be the default case unless the line was written in the log before `java.pm` execution as we discussed.

https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/b00b1c936621ffad5e75b0dfab0d8110004400cd/tests/console/java.pm#L49

or let's change grep to be more java greedy, 
`grep "java.*scriptlet failed"` 

It should not last there much. I hope that it will get fixed earlier if the problem will be highlighted.

Comment (Vogtinator) [reply tests/console/java.pm]: I think we're misunderstanding each other...

I'll try to explain my concern in a different way. Currently this does something like:
`zypper_call.retcode == 107 && zypper_call.failures.contains(java-9-javadoc scriptlet failed)`
while it should be
`zypper_call.retcode == 107 && zypper_call.failures == [java-9-javadoc scriptlet failed]`

IMO this case is too rare and the softfail was not necessary until now either, so I'd just drop it.

**Comment by dzedro** [comment]
Why is java-9-openjdk back in 15sp2 when it was gone in 12*, 15 and 15sp1 ?

File: tests/console/java.pm (hunk 2 above)
Comment (mloviska) [reply tests/console/java.pm]: > I think we're misunderstanding each other...
> 
> I'll try to explain my concern in a different way. Currently this does something like:
> `zypper_call.retcode == 107 && zypper_call.failures.contains(java-9-javadoc scriptlet failed)`
> while it should be
> `zypper_call.retcode == 107 && zypper_call.failures == [java-9-javadoc scriptlet failed]`
> 
> IMO this case is too rare and the softfail was not necessary until now either, so I'd just drop it.

But isn't the problem that the image pulls java-9 in fact? Because java-9 is not pulled by sle12sp{4,5} even. 
Checking scriptlet failures is not the proper detection, I admit it.

Comment (Vogtinator) [reply tests/console/java.pm]: The image does not pull it, the openQA test does.

**Comment by Vogtinator** [comment]
> Why is java-9-openjdk back in 15sp2 when it was gone in 12*, 15 and 15sp1 ?

This issue appears on Leap 15.1 but not 15.2. I assume it's the same for SLE 15 SP{1,2}.

**Comment by mloviska** [comment]
> > Why is java-9-openjdk back in 15sp2 when it was gone in 12*, 15 and 15sp1 ?
> 
> This issue appears on Leap 15.1 but not 15.2. I assume it's the same for SLE 15 SP{1,2}.

java-9-openjdk appears on [opensuse-15.2-JeOS-for-kvm-and-xen-x86_64](https://openqa.opensuse.org/tests/1206148#step/java/21) or https://openqa.opensuse.org/tests/1206148#step/java/16

File: tests/console/java.pm [hunk 4]
```diff
@@ -21,56 +21,35 @@ use base "consoletest";
...
-        }
+    if (script_run 'rpm -q wget') {
+        zypper_call 'in wget';
     }
-
-    zypper_call 'in wget';
     assert_script_run 'wget --quiet ' . data_url('console/test_java.sh');
     assert_script_run 'chmod +x test_java.sh';
-    assert_script_run './test_java.sh';
+    assert_script_run qq{./test_java.sh | tee -a /dev/$serialdev};
```
Comment (Vogtinator) [inline tests/console/java.pm]: Is `pipefail` set in this shell? Otherwise this won't detect errors

Comment (mloviska) [reply tests/console/java.pm]: No, it is handled in `script_output` to mine knowledge. Good that you have noticed it, thanks. 
Regarding sle '15+' and java testing. QAM enables all repositories in their tests, therefore their test suites pull java from legacy as well. In JeOS sle15sp2/sle15sp2 we do not really test the legacy module. 
Since the test case purpose is to install multiple java versions, I am about to enable the legacy module in `jeos-base+sdk` test suite or add legacy module registration/deregistration directly in the test case.

File: tests/console/java.pm [hunk 5]
```diff
@@ -21,56 +21,41 @@ use base "consoletest";
...
+
+    if (script_run 'rpm -q wget') {
+        zypper_call 'in wget';
     }
-
-    zypper_call 'in wget';
     assert_script_run 'wget --quiet ' . data_url('console/test_java.sh');
     assert_script_run 'chmod +x test_java.sh';
-    assert_script_run './test_java.sh';
+    my $pipe_fail = testapi::is_serial_terminal() ? 'bash -oe pipefail;' : 'bash -eox pipefail;';
```
Comment (Vogtinator) [inline tests/console/java.pm]: With the ';' it'll just start a new shell and wait for input definitely.

Comment (Vogtinator) [reply tests/console/java.pm]: In consoletest_setup, it already does `echo "set -o pipefail" >> /etc/bash.bashrc.local`, that might be enough?

Comment (mloviska) [reply tests/console/java.pm]: ah, I thought it just separates commands. thanks for hint

Comment (Vogtinator) [reply tests/console/java.pm]: Yes, exactly. First it starts `bash -oe pipefail`, which waits for input and after that exits, the parent shell runs the script.

Comment (mloviska) [reply tests/console/java.pm]: It should be enough as long as the test suite does not reboot VM.

```
    script_run 'echo "set -o pipefail" >> /etc/bash.bashrc.local';
    script_run '. /etc/bash.bashrc.local';
```

I haven't found ` /etc/bash.bashrc.local` in bash man though. Only `/etc/bash.bashrc ` ...

Comment (Vogtinator) [reply tests/console/java.pm]: > I haven't found /etc/bash.bashrc.local in bash man though. Only /etc/bash.bashrc

That's sourced by `/etc/bash.bashrc`. So as long as the serial terminal is opened after `consoletest_setup` applied the change, pipefail should be active.
